src/Mvc/Mvc.Core/src/Routing/UrlHelperBase.cs

```diff
@@ -43,45 +43,43 @@ protected UrlHelperBase(ActionContext actionContext)
         /// <inheritdoc />
         public virtual bool IsLocalUrl(string url)
         {
-            if (string.IsNullOrEmpty(url))
+            ReadOnlySpan<char> span = url;
+
+            if (span.IsEmpty)
             {
                 return false;
             }
 
-            // Allows "/" or "/foo" but not "//" or "/\".
-            if (url[0] == '/')
+            if (span[0] == '~')
             {
-                // url is exactly "/"
-                if (url.Length == 1)
-                {
-                    return true;
-                }
-
-                // url doesn't start with "//" or "/\"
-                if (url[1] != '/' && url[1] != '\\')
-                {
-                    return true;
-                }
+                span = span[1..];
+            }
 
+            // url does not start with "/" or "~/"
+            if (span[0] != '/')
+            {
                 return false;
             }
 
-            // Allows "~/" or "~/foo" but not "~//" or "~/\".
-            if (url[0] == '~' && url.Length > 1 && url[1] == '/')
+            // url is exactly "/" or "~/"
+            if ((uint)span.Length <= 1)     // PERF: elides the bounds-check below
             {
-                // url is exactly "~/"
-                if (url.Length == 2)
-                {
-                    return true;
-                }
+                return true;
+            }
 
-                // url doesn't start with "~//" or "~/\"
-                if (url[2] != '/' && url[2] != '\\')
+            // url doesn't start with "//" or "/\"
+            if (span[1] != '/' && span[1] != '\\')
+            {
+                // check that url does not contain any (ASCII) control character
+                for (var i = 1; i < span.Length; ++i)
```

Comment:
This could be vectorized quite straightforward. If it is worth it, I'll pick https://github.com/gfoidl/AspNetCore/compare/urlhelperbase-control-char...urlhelperbase-control-char-vectorized into this PR.

Maybe this method should also be `internal` so it could be used in more places?

Reply:
This method shouldn't be on any reasonable hot path. I'd just do something trivial that's easy to follow

Reply:
👍  I see vectors everywhere 😉